usr.sbin/ctld/ctld.cc

```diff
@@ -654,626 +562,1113 @@ parse_addr_port(char *arg, const char *def_port, struct addrinfo **ai)
 
-	if (TAILQ_EMPTY(&conf->conf_targets) ||
-	    TAILQ_EMPTY(&conf->conf_portal_groups))
-		return;
-	set_timeout(conf->conf_isns_timeout, false);
-	s = isns_do_connect(isns);
-	if (s < 0) {
-		set_timeout(0, false);
-		return;
+bool
+portal_group::set_offload(const char *offload)
+{
+	if (!pg_offload.empty()) {
+		log_warnx("cannot set offload to \"%s\" for "
+		    "portal-group \"%s\"; already defined",
+		    offload, name());
+		return (false);
 	}
-	error = gethostname(hostname, sizeof(hostname));
-	if (error != 0)
-		log_err(1, "gethostname");
 
-	if (oldisns == NULL || TAILQ_EMPTY(&oldisns->i_conf->conf_targets))
-		oldisns = isns;
-	isns_do_deregister(oldisns, s, hostname);
-	isns_do_register(isns, s, hostname);
-	close(s);
-	set_timeout(0, false);
+	pg_offload = offload;
+	return (true);
 }
 
-void
-isns_check(struct isns *isns)
+bool
+portal_group::set_pcp(u_int pcp)
 {
-	struct conf *conf = isns->i_conf;
-	int error, s, res;
-	char hostname[256];
+	if (pcp > 7) {
+		log_warnx("invalid PCP value %u for portal-group \"%s\"",
+		    pcp, name());
+		return (false);
+	}
 
-	if (TAILQ_EMPTY(&conf->conf_targets) ||
-	    TAILQ_EMPTY(&conf->conf_portal_groups))
-		return;
-	set_timeout(conf->conf_isns_timeout, false);
-	s = isns_do_connect(isns);
-	if (s < 0) {
-		set_timeout(0, false);
+	pg_pcp = pcp;
+	return (true);
+}
+
+bool
+portal_group::set_redirection(const char *addr)
+{
+	if (!pg_redirection.empty()) {
+		log_warnx("cannot set redirection to \"%s\" for "
+		    "portal-group \"%s\"; already defined",
+		    addr, name());
+		return (false);
+	}
+
+	pg_redirection = addr;
+	return (true);
+}
+
+void
+portal_group::set_tag(uint16_t tag)
+{
+	pg_tag = tag;
+}
+
+void
+portal_group::verify(struct conf *conf)
+{
+	if (pg_discovery_auth_group == nullptr) {
+		pg_discovery_auth_group = conf->find_auth_group("default");
+		assert(pg_discovery_auth_group != nullptr);
+	}
+
+	if (pg_discovery_filter == PG_FILTER_UNKNOWN)
+		pg_discovery_filter = PG_FILTER_NONE;
+
+	if (!pg_redirection.empty()) {
+		if (!pg_ports.empty()) {
+			log_debugx("portal-group \"%s\" assigned to target, "
+			    "but configured for redirection", name());
+		}
+		pg_assigned = true;
+	} else if (!pg_ports.empty()) {
+		pg_assigned = true;
+	} else {
+		if (pg_name == "default")
+			log_warnx("portal-group \"%s\" not assigned "
+			    "to any target", name());
+		pg_assigned = false;
+	}
+}
+
+/*
+ * Try to reuse a socket for 'newp' from an existing socket in one of
+ * our portals.
+ */
+bool
+portal_group::reuse_socket(struct portal &newp)
+{
+	for (portal_up &portal : pg_portals) {
+		if (newp.reuse_socket(*portal))
+			return (true);
+	}
+	return (false);
+}
+
+int
+portal_group::open_sockets(struct conf &oldconf)
+{
+	int cumulated_error = 0;
+
+	if (pg_foreign)
+		return (0);
+
+	if (!pg_assigned) {
+		log_debugx("not listening on portal-group \"%s\", "
+		    "not assigned to any target", name());
+		return (0);
+	}
+
+	for (portal_up &portal : pg_portals) {
```

Comment:
When the type is "simple" I prefer using the real type for ranged-for.  If the type is complicated (e.g. iterating over a map which uses an iterator type), I prefer to use `auto`.